suite/tests/client-interface/drreg-test.c

```diff
@@ -1254,6 +1463,414 @@ GLOBAL_LABEL(FUNCNAME:)
+         * Note that we do not respill aflags to the same slot, but select
+         * a new slot at each re-spill, so the app2app phase slot gets
+         * recycled and used by the insertion phase slot to re-spill the app
+         * aflags.
+         */
+#define FUNCNAME test_asm_faultN
+        DECLARE_FUNC_SEH(FUNCNAME)
+GLOBAL_LABEL(FUNCNAME:)
+#ifdef X86
+        PUSH_CALLEE_SAVED_REGS()
+        sub      REG_XSP, FRAME_PADDING /* align */
+        END_PROLOG
+
+        jmp      test23
+     test23:
+        mov      TEST_REG_ASM, DRREG_TEST_23_ASM
+        mov      TEST_REG_ASM, DRREG_TEST_23_ASM
+        mov      ah, DRREG_TEST_AFLAGS_ASM
+        sahf
+
+        /* app2app phase will reserve aflags here. */
+        mov      TEST_REG2_ASM, 1
+        /* insertion phase will reserve aflags here. */
+        mov      TEST_REG2_ASM, 2
+        /* app2app phase will release aflags here. */
+        mov      TEST_REG2_ASM, 3
+
+        mov      REG_XAX, 0
+        mov      REG_XAX, PTRSZ [REG_XAX] /* crash */
+        /* insertion phase will release aflags here. */
+        jmp      epilog23
+     epilog23:
+        add      REG_XSP, FRAME_PADDING /* make a legal SEH64 epilog */
+        POP_CALLEE_SAVED_REGS()
+        ret
+#elif defined(ARM)
+        /* XXX i#3289: prologue missing */
+        b        test23
+     test23:
+        movw     TEST_REG_ASM, DRREG_TEST_23_ASM
+        movw     TEST_REG_ASM, DRREG_TEST_23_ASM
+        msr      APSR_nzcvq, DRREG_TEST_AFLAGS_ASM
+
+        movw     TEST_REG2_ASM, 1
+        movw     TEST_REG2_ASM, 2
+        movw     TEST_REG2_ASM, 3
+
+        mov      r0, HEX(0)
+        ldr      r0, PTRSZ [r0] /* crash */
+
+        b        epilog23
+    epilog23:
+        bx       lr
+#elif defined(AARCH64)
+        /* XXX i#3289: prologue missing */
+        b        test23
+     test23:
+        movz     TEST_REG_ASM, DRREG_TEST_23_ASM
+        movz     TEST_REG_ASM, DRREG_TEST_23_ASM
+        movz     TEST_REG2_ASM, DRREG_TEST_AFLAGS_H_ASM, LSL 16
+        msr      nzcv, TEST_REG2_ASM
+
+        movz     TEST_REG2_ASM, 1
+        movz     TEST_REG2_ASM, 2
+        movz     TEST_REG2_ASM, 3
+
+        mov      x0, HEX(0)
+        ldr      x0, PTRSZ [x0] /* crash */
+
+        b        epilog23
+    epilog23:
+        ret
+#endif
+        END_FUNC(FUNCNAME)
+#undef FUNCNAME
+
+        /* Test 24: restore on fault for aflags restored once (for app read)
+         * before crash.
+         */
+#define FUNCNAME test_asm_faultO
+        DECLARE_FUNC_SEH(FUNCNAME)
+GLOBAL_LABEL(FUNCNAME:)
+#ifdef X86
+        PUSH_CALLEE_SAVED_REGS()
+        sub      REG_XSP, FRAME_PADDING /* align */
+        END_PROLOG
+
+        jmp      test24
+     test24:
+        mov      TEST_REG_ASM, DRREG_TEST_24_ASM
+        mov      TEST_REG_ASM, DRREG_TEST_24_ASM
+        mov      ah, DRREG_TEST_AFLAGS_ASM
+        sahf
+
+        mov      TEST_REG2_ASM, 1
+        /* Read aflags so that it is restored once. */
+        lahf
```

Comment:
Instead of the app using LAHF and SAHF in all of these tests, it seems better to use instructions that read/write OF as well in at least some of these tests, to exercise that code path?

Reply:
Done.